Fixed #16117 -- Added decorators for admin action and display functions.

pope1ni opened the pull request:
See ticket-16117. Also for reference, ticket-25134 and ticket-32099.

~~Opening as a draft as I still have work to do on the documentation - I just scattered a few changes in various areas to identify those areas that need work. Please feel free to review for the rest.~~

For some further detail:

I've added two decorators - one for custom display functions used in ``list_display`` and ``readonly_fields``, and one for custom action functions. Below are the proposed signatures of these decorators and how they map to the attributes assigned directly to the functions. Note that the attribute naming isn't copied to the decorator parameters exactly as they are often unnecessarily complex, e.g. `short_description` which happens to be the only description. The decorator parameters are also keyword-only which should prevent any mistakes in ordering of arguments.

----

```python
@admin.action(*, permissions=None, description=None)
```

| Decorator Parameter | Attribute |
| --- | --- |
| ``permissions`` | ``allowed_permissions`` |
| ``description`` | ``short_description`` |

```python
@admin.display(*, boolean=None, ordering=None, description=None, empty_value=None)
```

| Decorator Parameter | Attribute |
| --- | --- |
| ``boolean`` | ``boolean`` |
| ``ordering`` | ``admin_order_field`` |
| ``description`` | ``short_description`` |
| ``empty_value`` | ``empty_value_display`` |

----

It is probably worth mentioning some of the benefits of these decorators:

- The available attributes are more discoverable via code completion of the decorator parameters.
- The attributes assigned to the function could have a typo and not be used correctly.
  - This seems particularly bad for the ``allowed_permissions`` attribute on an action function.
- The ``@property`` decorator can be used which wasn't possible when assigning attributes directly to the function:
    ```python
    @property
    @admin.display(boolean=True)
    def is_published(self, obj):
        return self.date is not None
    ```

----

I also looked at some of the other things mentioned on the tickets and in existing third-party packages, e.g. [django-admin-decorators](https://github.com/kmike/django-admin-decorators). Here are my thoughts on various bits:

- Obviously there is no support for the - [now removed](https://code.djangoproject.com/ticket/25135#comment:10) - `allow_tags` attribute.
- I think that combined decorators are better than individual, e.g. `@boolean`, `@short_description`.
- These new decorators will not be mandatory and defining the attributes directly will still work. This avoids undue burden when upgrading. We'd also still need to write some attributes to the function, so there is no point making them private and going through a whole deprecation process.
- If none of the attributes are needed we also don't want to have to force use of `@admin.display()` with no arguments, although that is permissible to make it clear what that method is for. It'll just have no effect.
- Based on [this comment](https://code.djangoproject.com/ticket/25134#comment:1) I chose to use `ordering` instead of `admin_order_field` in the decorator. This is also because we now support expressions, so the `admin_order_field` name hasn't aged so well.
- Referring to [this comment](https://code.djangoproject.com/ticket/25134#comment:3), I added support for `empty_value_display` as that is already exposed in the same way as the other attributes. There is no existing support for `row_classes`, so I've ignored that.
- There was [this comment](https://code.djangoproject.com/ticket/25134#comment:4) regarding importing `django.contrib.admin` in a `models.py`. I don't necessarily think this is a problem if people desire to do this, but we are also not going to make using the decorator a hard requirement.
- Looking at what is provided by [django-admin-decorators](https://github.com/kmike/django-admin-decorators/blob/master/admin_decorators/__init__.py), we handle `@short_description`, `@order_field`, `@boolean`, ~~`@allow_tags`~~, and `@boolean`. There are also some other decorators, e.g. `@limit_width`, but these are manipulating the returned values which is out of the scope.

----

Questions for those reviewing:

- ~~Are we happy with the decorator names?~~ (Dropped the `_method` suffix.)
  - ~~The `_method` suffix is so-so as functions can be used instead of `ModelAdmin` methods.~~
- Are we happy with the decorator parameter names?

AstroMatt commented:
I have created this request 9 years ago but this is still valid issue for me :)
Thanks! :)

pope1ni commented:
Ok. This is ready for review.

adamchainz commented on django/contrib/admin/actions.py:
```diff
@@ -4,12 +4,14 @@
 from django.utils.translation import gettext as _, gettext_lazy
 
 
+@action_method(permissions=['delete'], description=gettext_lazy("Delete selected %(verbose_name_plural)s"))
```
Line break for future black compat? 😇

```suggestion
@action_method(
    permissions=['delete'],
    description=gettext_lazy("Delete selected %(verbose_name_plural)s"),
)
```

carltongibson replied:
Black won't mind. 😀

pope1ni replied:
I guess the description is quite long, so probably worth breaking the line.

adamchainz commented on django/contrib/admin/decorators.py:
```diff
@@ -1,3 +1,63 @@
+        make_published.allowed_permissions = ['publish']
+        make_published.short_description = 'Mark selected stories as published'
+    """
+    def decorator(f):
```
Not sure the style guide says it but I prefer variable names >1 character when they're used across multiple lines, so `func` here?

pope1ni replied:
Fair enough.

adamchainz commented on django/contrib/admin/decorators.py:
```diff
@@ -1,3 +1,63 @@
+        def make_published(self, request, queryset):
+            queryset.update(status='p')
+
+    This is equivalent to setting the attributes on the method directly::
```
It's a little confusing to not signpost the mismatch between the attribute names. Especially the wording "the attributes" sounds like they should be the same. Perhaps

```suggestion
    This is equivalent to setting some attributes (with the original, longer names) on the function directly::
```

pope1ni replied:
That's nicer. And I'll change to generally prefer function throughout as methods are functions bound to an object.

adamchainz commented on django/contrib/admin/decorators.py:
```diff
@@ -1,3 +1,63 @@
+def action_method(*, permissions=None, description=None):
+    """
+    Conveniently add attributes to an action method::
```
```suggestion
    Conveniently add attributes to an action function::
```

adamchainz commented on django/contrib/admin/decorators.py:
```diff
@@ -1,3 +1,63 @@
+        def is_published(self, obj):
+            return obj.publish_date is not None
+
+    This is equivalent to setting the attributes on the method directly::
```
Ditto above improved wording

adamchainz commented on django/contrib/admin/decorators.py:
```diff
@@ -1,3 +1,63 @@
+
+def display_method(*, boolean=None, ordering=None, description=None, empty_value=None):
+    """
+    Conveniently add attributes to a display method::
```
```suggestion
    Conveniently add attributes to a display function::
```

adamchainz commented on django/contrib/admin/options.py:
```diff
@@ -869,7 +870,7 @@ def _get_base_actions(self):
             if name in base_action_names:
                 continue
-            description = getattr(func, 'short_description', name.replace('_', ' '))
+            description = getattr(func, 'short_description', capfirst(name.replace('_', ' ')))
```
Slightly unrelated/undiscussed change?

pope1ni replied:
That's why it's in a separate commit with the message `Refs #16117 -- Standardized admin action description generation.` The default name for the actions defined in `ModelAdmin.actions` uses `capfirst()`, but the default name for those defined in `AdminSite.actions` doesn't.

Sorry. There are a few cleanup commits in here where I found some warts while searching for all of the code related to these attributes.

adamchainz reviewed (commented):
❤ Thanks for doing this! Much nicer, and more type-hint-able in the future.

> The _method suffix is so-so as functions can be used instead of ModelAdmin methods.

I feel like dropping the suffix would be a good idea. `@action` reads better and avoids confusion around their validity for plain functions.

pope1ni commented:
> > The _method suffix is so-so as functions can be used instead of ModelAdmin methods.
> 
> I feel like dropping the suffix would be a good idea. `@action` reads better and avoids confusion around their validity for plain functions.

That works for me. In the general case if people are doing `from django.contrib import admin`, then they'll use `@admin.action()` and `@admin.display()`.

pope1ni commented:
Thanks for the initial review @adamchainz.

adamchainz commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -395,12 +402,38 @@ For example::
+
+    .. versionadded:: 3.2
+
+    This decorator can be used for setting specific properties on custom
```
```suggestion
    This decorator can be used for setting specific attributes on custom
```

adamchainz commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -3290,6 +3292,38 @@ The action in the examples above match the last part of the URL names for
+
+    .. versionadded:: 3.2
+
+    This decorator can be used for setting specific properties on custom
```
```suggestion
    This decorator can be used for setting specific attributes on custom
```

adamchainz commented on docs/topics/i18n/translation.txt:
```diff
@@ -405,6 +407,15 @@ with the ``short_description`` attribute::
+Note that this is also applicable if setting the ``short_description``
+attribute on the method directly instead of using the decorator::
+
+        ...
```
I think this example would be clearer by showing the container:

```suggestion
    class MyThing(models.Model):
        ...
```

But I also wonder if it's still required - the full docs for `@display` show the equivalent direct attribute setting, so it seems redundant to re-mention it here.

pope1ni replied:
Yes. I wasn't originally sure, but I think we can drop it.

adamchainz reviewed (commented):
Some possible small docs tweaks

pope1ni commented:
Thanks again, @adamchainz.

I added a fixup that allows using the empty decorator without parentheses. I also added some documentation to note that use of the decorator is not compulsory, but can still be useful as a marker to identify the purpose of the function even where adding attributes is not required/desired.

I'm happy to add an extra wrapping function and use `functools.wraps()` if you would like this to avoid side effects @carltongibson. Just let me know... :stuck_out_tongue:

adamchainz commented:
IMO the side-effects issue is less of a worry here, since the existing method was to attach the extra attributes to the function directly.

pope1ni commented:
I've split out the clean up commits into separate pull requests and removed most of them from here. I'll remove the remaining two when the other pull requests are merged and I can rebase this one.

adamchainz commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -395,12 +402,48 @@ For example::
+        def make_published(self, request, queryset):
+            queryset.update(status='p')
+        make_published.allowed_permissions = ['publish']
+        make_published.short_description = 'Mark selected stories as published'
```
So if I come to this documentation (or for `@display`) for the first time, it tells me about setting these attributes but not what they mean, and nor is there an obvious link to where to read about them. It makes it a bit hard to use `@action` or `@display` - it still feels like I have to know about both ways.

Perhaps the decorator reference should become the canonical place where the meanings of the arguments and attributes are documented? At least, they can be briefly described with links to "more info" sections below, with each of the examples. The current structure is a fairly messy since it seems to have grown organically.

For actions:

* `short_description` is under "Writing action functions" only
* `allowed_permissions` has a whole section "Setting permissions for actions"

For display, everything is listed in bullet points under the `ModelAdmin.list_display` documentation (a giant list that starts "A few special cases"):

* `boolean` - this unnecessarily re-spells what kinds of strings and functions are allowed in `list_display`
* `admin_order_field`
* `short_description` - this only has one sentence saying "As some examples have already demonstrated" which doesn't feel very easy to use as a reference since it's saying "go read above"
* `empty_value_display` - this goes off on a bit of a tangent documenting the ability to change it for the whole admin site and the model admin class too. It would probably be best as its own section that documentation for the other two classes link to.

I am suggesting a fair amount of docs rearrangement here. I am happy to shunt it to a second PR and help on that.

pope1ni replied:
I agree. It is quite impenetrable and definitely needs rejigging. I tried not to move stuff around too much at this stage, however, and kept everything in the same place.

I think a follow up PR to improve the whole of this section would be a good idea - the large bulleted list doesn't feel the best way to present the options.

adamchainz reviewed (commented):
Looking good.. mainly I think we could improve the docs here

pope1ni commented on django/contrib/admin/decorators.py:
```diff
@@ -1,3 +1,71 @@
+def action(function=None, *, permissions=None, description=None):
```
> Was just thinking through what this means for ordering of arguments. `function` is a defaulted argument that may be passed positionally _or_ by keyword. So `action(permissions=..., function=...)` worked.

While true, this is intended to be used as a decorator, so the wrapped function will always be passed as the first positional argument. If you try to pass `function=...` to the decorator, it'll fail as the remaining arguments are keyword only. This is also why I didn't mention `function` in the signatures in the documentation, as this is really just an implementation detail to avoid needing to have empty parentheses if no arguments are provided.

> ...making `function` positional only with a default, but that's only available on Python 3.8+.

Actually, I think we'd want `def action(function=None, /, *, permissions=None, description=None):` to enforce that `function` is positional-only and `description` and `permissions` are keyword-only. I think there is benefit to them being keyword only, as it forces them to be explicitly defined. This is particularly important where permissions are concerned. But as you say, waiting for 3.8.

adamchainz replied:
> Actually, I think we'd want def action(function=None, /, *, permissions=None, description=None): to enforce that function is positional-only and description and permissions are keyword-only.

Ah yes, my bad! Gosh these separators are a bit tricky :)

AstroMatt replied:
Here is the explanation I wrote some time ago :)
https://python.astrotech.io/advanced/function/parameter-syntax.html

pope1ni commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -3290,6 +3293,48 @@ The action in the examples above match the last part of the URL names for
+        is_published.admin_order_field = '-publish_date'
+        is_published.short_description = 'Is Published?'
+
+    Also note that the ``empty_value`` decorator parameter maps to the
+    ``empty_value_display`` attribute assigned directly to the function.
```
The example is using `boolean` which will display icons, so `empty_value` would be pointless, no? We either leave this sentence or we would need to add a second example using `empty_value`.

I can't think of a scenario where it makes sense to use them together - they're [mutually exclusive](https://github.com/django/django/blob/c70cd2a926ffab47f6613e83e0c8828eb6c2c064/django/contrib/admin/utils.py#L414-L417). Maybe we should warn people if they do? Or not allow it (at least via the decorator)?

adamchainz replied:
Good point. Yes, disallowing in the decorator would make sense to me - whilst we're building a new API we might as well help people out here.

pope1ni replied:
Thanks Adam. Have pushed a fixup commit to handle this case.

pope1ni commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -754,32 +751,38 @@ subclass::
           class PersonAdmin(admin.ModelAdmin):
               list_display = ('full_name',)
 
+      Please note, however, that the ``property()`` function and **not** the
```
Thanks. I've tweaked as suggested.

adamchainz reviewed (approved):
Alright LGTM, although I'm sure the fellows will have some points to check

![](https://media.giphy.com/media/3o6Zt8GgkGACnMCJoc/source.gif)

pope1ni commented:
@carltongibson Any chance this will make it in for 3.2? It's been "ready" for quite a while.

carltongibson commented:
> I'm happy to add an extra wrapping function and use functools.wraps() if you would like this to avoid side effects @carltongibson. Just let me know... 😛

Side-effects matter, for example, when you use a decorator inline in the URL conf, like this... 

```
urlpatterns = [
     path("decorated-stuff/", decorate(stuff), name="decorated-stuff"), 
]
```

This is pretty common. Then, it's **surprising**, shall we say, if `stuff` has additional attributes.

carltongibson commented on django/contrib/admin/actions.py:
```diff
@@ -4,12 +4,17 @@
 
+@action(
+    permissions=['delete'],
+    description=gettext_lazy("Delete selected %(verbose_name_plural)s"),
```
```suggestion
    description=gettext_lazy('Delete selected %(verbose_name_plural)s'),
```

pope1ni commented:
> Side-effects matter, for example, when you use a decorator inline in the URL conf, like this...

@carltongibson I've pushed a `!fixup` to prevent side effects.

carltongibson reviewed (commented):
Hey @pope1ni — thanks for this. Nice work on a 10 year old ticket. 😀

I really like it. 

I can accept that a bigger docs re-write can wait for a second PR. (But I'm thinking we shouldn't leave that forever.)

I have **one concern**. We have thousands upon thousands of sites out there using the old attributes, who (I'm imagining) might want to search for the attributes they're using in the docs to lookup how they work and what they do. 

I'm a **bit worried** that we make such too difficult removing all but the single reference in the "with the original, longer names" bit of the added decorator docs. 

Related, I'm not seeing `versionchanged`/`versionadded` annotations. 

I think we probably need to accommodate developers working on existing (legacy) installs for some time to come. Maybe a prominent link to the relevant decorator at the beginning of the actions and list_display sections? 🤔 What do you think? 

Other than that I don't see why we can't have it. 
Nice. 👍

carltongibson commented:
>  I've pushed a !fixup to prevent side effects.

Hey @pope1ni — I'm not sure I intended that. 😄 The way you'd use **these** decorators is **replacing** the function in place, so no harm adjusting the original. (Similar to [Adam's comment above](https://github.com/django/django/pull/13532#issuecomment-718082554) — the old way adds attributes to the method directly too...)

pope1ni commented:
> The way you'd use **these** decorators is **replacing** the function in place, so no harm adjusting the original.

That is true. I guess the changes made don't harm things other than having extra function call overhead. I'll leave the `!fixup` as it is and it can be dropped when you squash.

pope1ni commented:
> I have **one concern**. We have thousands upon thousands of sites out there using the old attributes, who (I'm imagining) might want to search for the attributes they're using in the docs to lookup how they work and what they do.
> 
> I'm a **bit worried** that we make such too difficult removing all but the single reference in the "with the original, longer names" bit of the added decorator docs.

I can sympathise with that. I felt the same but wasn't sure how to address it.

> Related, I'm not seeing `versionchanged`/`versionadded` annotations.

I already had `versionadded` annotations in the documentation of the generators.

I've pushed another `!fixup` with some `versionchanged` annotations in the relevant areas. Maybe the wording needs tweaking, but it's a start... I figured that this would be enough for mentioning the original attribute names for a couple of releases.

> I think we probably need to accommodate developers working on existing (legacy) installs for some time to come. Maybe a prominent link to the relevant decorator at the beginning of the actions and list_display sections?  What do you think?

I had a look at this, but wasn't sure how best to fit anything in as the documentation is currently structured. For the `list_display` section it isn't really relevant until it starts talking about callables.

> Nice.

:smiley:

carltongibson commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -551,7 +558,9 @@ subclass::
     column that displays the ``__str__()`` representation of each object.
 
-    There are four types of values that can be used in ``list_display``:
+    There are four types of values that can be used in ``list_display``. All
+    but the simplest may use the  :func:`~django.contrib.admin.display`
+    decorator is used to customize how the field is presented:
```
In the end, I just added the additional link to the decorator here. Reading through I think that's enough.

carltongibson reviewed (approved):
Hey @pope1ni — I went over the docs again, I think they're OK, I made the single adjustment. 

Thanks for this! Let's risk it for a biscuit. 😀

pope1ni commented:
> Let's risk it for a biscuit.

:cookie:

pope1ni commented:
Thanks @adamchainz and @carltongibson! :superhero:

adamchainz commented:
![](https://media.giphy.com/media/l4hLHJCwpscYLkArm/source.gif)
